lib/SILOptimizer/Transforms/RetainReleaseCodeMotion.cpp

```diff
@@ -0,0 +1,1070 @@
+  /// or release of the refcounted root to be moved across.
+  llvm::SmallBitVector BBKillSet;
+
+  /// A bit vector for which the ith bit represents the ith refcounted root in
+  /// RCRootVault. If this bit is set, that means this is potentially a retain
+  /// or release that can be sunk or hoisted to this point. This is used to
+  /// optimize the time for computing genset and killset.
+  /// NOTE: this vector contains an approximation of whether there will be a
+  /// retain or release to a certain point of a basic block.
+  llvm::SmallBitVector BBMaxSet;
+};
+
+/// CodeMotionContext - This is the base class which retain code motion and
+/// release code motion inherits from. It defines an interface as to how the
+/// code motion precedure should be.
+class CodeMotionContext {
+protected:
+  /// Dataflow needs multiple iteration to converge. If this is false, then we
+  /// do not need to generate the genset or killset, i.e. we can simply do 1
+  /// pessimistic data flow iteration.
+  bool MultiIteration;
+
+  /// The allocator we are currently using.
+  llvm::BumpPtrAllocator &BPA;
+
+  /// Current function we are analyzing.
+  SILFunction *F;
+
+  /// Current post-order we are using.
+  PostOrderFunctionInfo *PO;
+
+  /// Current alias analysis we are using.
+  AliasAnalysis *AA;
+
+  /// Current rc-identity we are using.
+  RCIdentityFunctionInfo *RCFI;
+
+  /// All the unique refcount roots retained or released in the function.
+  llvm::SetVector<SILValue> RCRootVault;
+
+  /// Contains a map between rc roots to their index in the RCRootVault.
+  /// used to facilitate fast rc roots to index lookup.
+  llvm::DenseMap<SILValue, unsigned> RCRootIndex;
+
+  /// All the retains or releases originally in the function. Eventually
+  /// they will all be removed after all the new ones are generated.
+  llvm::DenseSet<SILInstruction *> RCInstructions;
+
+  /// All the places to place the new retains or releases after code motion.
+  using InsertPointList = llvm::SmallVector<SILInstruction *, 2>;
+  llvm::SmallDenseMap<SILValue, InsertPointList> InsertPoints;
+
+  /// These are the blocks that have a RC instruction to process or it blocks
+  /// some RC instructions. If the basic block has neither, we do not need to
+  /// process the block again in the last iteration. We populate this set when
+  /// we compute the genset and killset.
+  llvm::DenseSet<SILBasicBlock *> InterestBlocks;
+
+  /// An RC-identity cache.
+  /// TODO: this should be cached in RCIdentity analysis.
+  llvm::DenseMap<SILValue, SILValue> RCCache;
+
+  /// Return the rc-identity root of the SILValue.
+  SILValue getRCRoot(SILValue R) {
+    auto Iter = RCCache.find(R);
+    if (Iter != RCCache.end())
+      return Iter->second;
+     return RCCache[R] = RCFI->getRCIdentityRoot(R);
+  }
+
+  /// Return the rc-identity root of the instruction.
+  SILValue getRCRoot(SILInstruction *I) {
+    return getRCRoot(I->getOperand(0));
```

Comment:
This should be asserting on the kinds of instructions that are valid here. It's clear from the uses of the function that it's only called on certain instructions, but it should be made evident in the function body as well.